### PR TITLE
feat: add last value caching

### PR DESCRIPTION
**Pull request by dignifiedquire**
Experiment to improve message loading in orbit.

https://github.com/ipfs/pm/issues/355

cc @haadcode

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -53,6 +53,13 @@ class FloodSub extends EventEmitter {
      */
     this.subscriptions = new Set()
 
+    /**
+     * Cache for the last message on seen subscriptions
+     * @type {Map<string, Buffer>}
+     * @private
+     */
+    this._lvc = new Map()
```
Comment (Kubuxu) [inline src/index.js]: I would suggest using LRU cache, caches like that bitten as so many times in go-ipfs as things scaled up.

Comment (dignifiedquire) [inline src/index.js]: Yeah need to figure out what's the best caching strategy here. For now I just want to test if this actually solves our problems

Comment (dignifiedquire) [inline src/index.js]: done @Kubuxu better safe than sorry :)

**Comment by diasdavid** [comment]
Note that this is a change in the spec. Good to try new things, but remember that this won't be present in go-ipfs unless it gets pushed to the spec and implemented in go-ipfs too. //cc @whyrusleeping 

This last message cache is something super biased for Orbit. What we are looking for is a SACK (Selective Ack) strategy, so that new peers can NACK the last message.

**Comment by dignifiedquire** [comment]
@diasdavid please see https://github.com/libp2p/research-pubsub/issues/12 for discussion on this

> This last message cache is something super biased for Orbit. What we are looking for is a SACK (Selective Ack) strategy, so that new peers can NACK the last message

could you explain how this would give us last message caching?

**Comment by dignifiedquire** [comment]
> Note that this is a change in the spec

yes, but the spec is not properly written down, which makes it hard to properly propose changes.

> but remember that this won't be present in go-ipfs unless it gets pushed to the spec and implemented in go-ipfs too

That is fine. It looks like this works really well, so once we have an agreement on the spec we can move to implementing it in go as well.

**Comment by dignifiedquire** [comment]
I also want to point out, that while I agree that spec updates are important, the whole floodsub story is still behind an experimental flag, with the goal to get something that works for the consumers of the api. This change is quite an important one, especially on the js side, as there is no ipns that could be used to do a similar thing. I also do not agree that it's fully tailored for Orbit. While that was the motivation it is a problem that many systems trying to propagate state through ipfs using pubsub will run into.

**Pull request closed without merging**